runtime/lua/vim/diagnostic.lua

```diff
@@ -813,11 +813,39 @@ local function next_diagnostic(position, search_forward, bufnr, opts, namespace)
   position[1] = position[1] - 1
   bufnr = get_bufnr(bufnr)
   local wrap = if_nil(opts.wrap, true)
-  local line_count = api.nvim_buf_line_count(bufnr)
+
+  local auto_severity = opts.severity == 'auto'
+  if auto_severity then
+    -- Clear the severity filter so that we get all diagnostics
+    opts.severity = nil
+  end
+
   local diagnostics =
     get_diagnostics(bufnr, vim.tbl_extend('keep', opts, { namespace = namespace }), true)
+
+  -- When 'auto' severity is set, sort the diagnostics by severity. The first diagnostic then
+  -- contains the highest severity, and we can discard all diagnostics with a lower severity.
+  if auto_severity then
+    table.sort(diagnostics, function(a, b)
+      return a.severity < b.severity
+    end)
+
+    -- Find the first diagnostic where the severity does not match the highest severity, and remove
+    -- that element and all subsequent elements from the array
+    local len = #diagnostics
+    for i = 2, len do
+      if diagnostics[i].severity ~= diagnostics[i - 1].severity then
```

Comment:
Could `diagnostics[i - 1].severity` be lifted out of the loop as `local worst = diagnostics[1].severity`, to avoid the lookup on each item?

Reply:
db177dc69170a346a177b522c857e242085dc8b7